### PR TITLE
ASoC: SOF: ipc4-control: Add support for bytes control type

### PR DESCRIPTION
**Pull request by ujfalusi**
Hi,

RFC state atm due to few uncertainty regarding to the `struct sof_ipc4_abi_hdr` and the definition of the blob_type number in there.
There is also the `header.numid` in the TLV (bytes_ext) which in ipc3 was basically was set to `SOF_CTRL_CMD_BINARY` for all binary controls, but it is not clear what it should be with IPC4 and how that is used. This brings up some questions regarding to validation of the blob, is it sent for the correct module, and is it generic identification so that it can be used on different systems (not runtime determined 'random' number). Perhaps we should have UUID in the ABI header for IPC4 blobs like these?

the series implements support for MOD_INIT_INSTANCE (blob_type = 0) and MOD_LARGE_CONFIG_SET (blob_type = 4) type of blobs and tested with the later using FIR module.

### CODE REVIEW

**Comment by ujfalusi** [comment]
Changes since v1:
- rebased on topic/sof-dev

**Review by singalsu — DISMISSED** [review]
Thanks Peter! This PR works perfectly with FIR EQ component's byte control set/get (patches needed for FIR, topology, rimage).

**Comment by ujfalusi** [comment]
Changes since v2:
- fixup patches to remove the `struct sof_ipc4_abi_hdr` and re-use/extend the existing `struct sof_abi_hdr` for IPC4 also.
- fixup to not store the binary blob under the widget as well since it should not be sent at mod_init_instance time.
- the binary control support is updated accordingly to the ABI header change.

@singalsu, @juimonen if you still have the blob you have tested with EQ then the only change needed for you is to move the size back where it is with IPC3 (3rd u32 instead the 2nd u32). Afaik the EQ module in firmware does not use the param_id, so leave the 2nd u32 as 0 to be safe (param_id = 0).

File: include/uapi/sound/sof/header.h [hunk 1]
```diff
@@ -25,35 +31,21 @@
...
+			__u8 reserved[3];
+		} ipc4;
+	};
 	__u32 size;
 	__u32 abi;
 	__u32 reserved[4];
 	__u32 data[];
 }  __packed;
 
-/**
```
Comment (ranj063) [inline include/uapi/sound/sof/header.h:29]: if we're removing this anyway, why bother with the first patch at all?

File: sound/soc/sof/ipc4-topology.c [hunk 2]
```diff
@@ -893,18 +890,6 @@ static int sof_ipc4_widget_setup_comp_process(struct snd_sof_widget *swidget)
 	if (process->payload_with_output_fmt)
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c]: @ujfalusi while you are at it. Can you please also fix this part? Looks like init instance with output format isnt supported either right? The wov module needs this and will just have to adapt to use the base_cfg_ext right?

File: sound/soc/sof/ipc4-topology.c [hunk 3]
```diff
@@ -1646,6 +1648,29 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
 		cfg += sizeof(struct sof_ipc4_audio_format);
 	}
 
+	fw_module = swidget->module_info;
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c]: do we need to do this in prepare? Can be done in setup no?

File: sound/soc/sof/ipc4-control.c [hunk 4]
```diff
@@ -194,19 +392,25 @@ static int sof_ipc4_widget_kcontrol_setup(struct snd_sof_dev *sdev, struct snd_s
...
-				if (ret < 0) {
-					dev_err(sdev->dev, "kcontrol %d set up failed for widget %s\n",
+				if (ret < 0)
+					dev_err(sdev->dev,
+						"kcontrol %d set up failed for widget %s\n",
+						scontrol->comp_id, swidget->widget->name);
+				break;
+			case SND_SOC_TPLG_CTL_BYTES:
+				ret = sof_ipc4_set_bytes_data(sdev, scontrol, false);
+				if (ret < 0)
```
Comment (ranj063) [inline sound/soc/sof/ipc4-control.c]: can the check go out side switch so we dont duplicate in both cases

File: sound/soc/sof/ipc4-control.c [hunk 5]
```diff
@@ -181,11 +181,209 @@ static int sof_ipc4_volume_get(struct snd_sof_control *scontrol,
...
+}
+
+static int sof_ipc4_bytes_ext_get(struct snd_sof_control *scontrol,
+				  const unsigned int __user *binary_data,
+				  unsigned int size)
+{
+	struct snd_ctl_tlv __user *tlvd = (struct snd_ctl_tlv __user *)binary_data;
+	struct sof_ipc4_control_data *cdata = scontrol->ipc_control_data;
+	struct snd_soc_component *scomp = scontrol->scomp;
+	struct sof_abi_hdr *data = cdata->data;
```
Comment (ranj063) [inline sound/soc/sof/ipc4-control.c:346]: what if a kcontrol has no data in topology? cdata->data will be NULL until the user does a kcontrol set isnt it? In that case, dont we actually need to do a GET_LARGE_CONFIG to fetch the data from the DSP? And for volatile kcontrols we should always fetch this data from the DSP right?

File: include/uapi/sound/sof/header.h (hunk 1 above)
Comment (ujfalusi) [reply include/uapi/sound/sof/header.h:29]: We are not removing it, but for upstream this will be the first patch, then the extending to IPC4.

File: sound/soc/sof/ipc4-topology.c (hunk 2 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.c]: Right, so another fixup to drop the whole token and flag?

File: sound/soc/sof/ipc4-topology.c (hunk 3 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.c]: In setup we do not have the control_data, I think, so we need to do this here (also other type do it in there prepare)

File: sound/soc/sof/ipc4-control.c (hunk 5 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-control.c:346]: I see, hrm, I see that we are fiddling with things in IPC3 side. Let me see how it can be done, but if the cdata->data is NULL then we do not have binary support in the first place, we need data to be allocated already.
If you mean that the content is 0, yeah, that might be the case, I have not implemented the volatile in this initial PR, but it is planed for later time.

File: sound/soc/sof/ipc4-control.c (hunk 4 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-control.c]: sure.

**Comment by ujfalusi** [comment]
Changes since v3:
- Drop payload_with_output_fmt and SOF_TKN_PROCESS_PAYLOAD_WITH_OUTPUT_FMT in fixups
- implement bytes_ext_volatile_get
- Move the error print outside of the switch to de-duplicate it in sof_ipc4_widget_kcontrol_setup()

File: sound/soc/sof/ipc4-control.c [hunk 6]
```diff
@@ -181,32 +181,260 @@ static int sof_ipc4_volume_get(struct snd_sof_control *scontrol,
...
+	 */
+	if (size < sizeof(struct snd_ctl_tlv))
+		return -ENOSPC;
+
+	size -= sizeof(struct snd_ctl_tlv);
+
+	/* get all the component data from DSP */
+	if (from_dsp) {
+		struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+		int ret = sof_ipc4_set_bytes_data(sdev, scontrol, true);
```
Comment (ranj063) [inline sound/soc/sof/ipc4-control.c]: is this correct @ujfalusi . Why do we set when we're doing a get?

File: sound/soc/sof/ipc4-topology.c [hunk 7]
```diff
@@ -1661,29 +1623,27 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
...
+		struct sof_ipc4_control_data *control_data;
+		struct snd_sof_control *scontrol;
+		struct soc_bytes_ext *sbe;
 
 		/* payload uses byte kcontrol */
 		if (widget->dobj.widget.kcontrol_type[i] != SND_SOC_TPLG_TYPE_BYTES)
 			continue;
+
 		sbe = (struct soc_bytes_ext *)kc->private_value;
 		scontrol = sbe->dobj.private;
```
Comment (aiChaoSONG) [inline sound/soc/sof/ipc4-topology.c]: We are looping the asoc widgets' kcontrol list and doing dereference directly, Pierre questions if it is safe before:https://github.com/thesofproject/linux/pull/4121#discussion_r1068580516, isn't it still a problem now?

**Review by aiChaoSONG — COMMENTED** [review]
I got a chance to test this PR for smart amp. we are able to send configuration blob with this pr.

File: sound/soc/sof/ipc4-topology.c (hunk 7 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.c]: If the kcontrol type is SND_SOC_TPLG_TYPE_BYTES we know exactly what is behind the private_value, but I will check if this can be made cleaner.

Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.c]: This part is now gone.

**Comment by ujfalusi** [comment]
Changes since v4:
- Drop the rest of the references to `SOF_PROCESS_TOKENS` in the process widget fixup to be able to probe when there is a process widget in topology
- New function to set the module_id in the message template for the kcontrols associated with a widget
- Use the new helper to set the module_id for the message template associated with a process  module (we could have multiple kcontrols!)
- Now the volatile get should be working (I have no means to test it), put is working.

File: sound/soc/sof/ipc4-topology.c [hunk 8]
```diff
@@ -1706,6 +1709,11 @@ static int sof_ipc4_control_load_bytes(struct snd_sof_dev *sdev, struct snd_sof_
 		}
 	}
 
+	msg = &control_data->msg;
+	msg->primary = SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_MOD_LARGE_CONFIG_SET);
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.c:1713]: this is modifed in sof_ipc4_set_get_data() to SET/GET depending on whether it is a put/get. So maybe we dont need to se this here?

Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.c:1713]: Yes, true. I copied the way how the volume control message is set up. It is easier to follow and understand by the reader and does not require knowledge that what will happen later on.

**Comment by singalsu** [comment]
This version works perfectly with FIR and IIR in topology in https://github.com/thesofproject/sof/pull/6970 .

File: sound/soc/sof/ipc4-topology.h [hunk 9]
```diff
@@ -349,12 +349,6 @@ struct sof_ipc4_process {
 	void *ipc_config_data;
 	uint32_t ipc_config_size;
 	struct sof_ipc4_msg msg;
-	/*
```
Comment (ranj063) [inline sound/soc/sof/ipc4-topology.h:352]: @ujfalusi you missed one thing here. I think we should also remove the output_format field in this struct right?

**Review by ranj063 — COMMENTED** [review]
Looks great @ujfalusi. Just one comment abotu removing the out_format from struct sof_ipc4_process

File: sound/soc/sof/ipc4-control.c [hunk 10]
```diff
@@ -181,32 +181,262 @@ static int sof_ipc4_volume_get(struct snd_sof_control *scontrol,
...
+	if (copy_from_user(data, tlvd->tlv, header.length))
+		return -EFAULT;
+
+	sof_ipc4_set_get_bytes_data(sdev, scontrol, true, true);
+
+	return 0;
+}
+
+static int _sof_ipc4_bytes_ext_get(struct snd_sof_control *scontrol,
+				  const unsigned int __user *binary_data,
```
Comment (ranj063) [inline sound/soc/sof/ipc4-control.c]: there's an alignment issue here. Why doesnt checkpatch report it?

File: sound/soc/sof/ipc4-topology.h (hunk 9 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.h:352]: Actually @ujfalusi please don't remove the output_format. I think we will need it.

**Review by ranj063 — DISMISSED** [review]

**Comment by ujfalusi** [comment]
Changes since v5:
- parameter alignment fix in the last patch for _sof_ipc4_bytes_ext_get() (catched by @ranj063 and checkpatch --strict)

**Review by juimonen — DISMISSED** [review]
looks good and logical series to me. Existing comments by previous reviewers that are not addressed, should be addressed/closed.

File: sound/soc/sof/ipc4-control.c [hunk 11]
```diff
@@ -181,32 +181,262 @@ static int sof_ipc4_volume_get(struct snd_sof_control *scontrol,
...
+	int ret = 0;
+
+	/* Send the new data to the firmware only if it is powered up */
+	if (set && !pm_runtime_active(sdev->dev))
+		return 0;
+
+	msg->extension = SOF_IPC4_MOD_EXT_MSG_PARAM_ID(data->ipc4.param_id);
+
+	msg->data_ptr = data->data;
+	msg->data_size = data->size;
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-control.c:200]: Hrm, interesting. We should be sending the sof_abi_hdr part as well to the firmware, not only the blob?

In IPC3 we have `struct sof_ipc_ctrl_data` which wraps all sorts of kcontrol related data (volume, enum, bytes), at the end of `struct sof_ipc_ctrl_data` we have a union and one option is the `sof_abi_hdr` (sof_ipc_ctrl_value_chan / sof_ipc_ctrl_value_comp / sof_abi_hdr)

With IPC4 we don't have such a generic kcontrol message, but we send everything with large_config and the target  module and it's parameter will tell the recipient what the data is and it will know how to handle it.

It is simple to send the header along, but it will need firmware side changes to handle the additional wrap (and btw, the abi_hdr use in firmware is just to `&cdata->data->data` to get to the real blob and the header's content ignored.

**Review by bardliao — DISMISSED** [review]

File: sound/soc/sof/ipc4-control.c (hunk 11 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-control.c:200]: Right, after a quick check in my brain: it does not make sense to send the sof_abi_hdr along with the blobs. It could be only understood by SOF modules and IPC4 is much broader, we can not start sending blobs with strange headers to existing modules.

File: include/uapi/sound/sof/header.h [hunk 12]
```diff
@@ -25,35 +31,21 @@
...
-	__u32 type;
+	union {
+		struct {
+			__u32 type;
+		} ipc3;
+		struct {
+			__u8 param_id;
+			__u8 reserved[3];
+		} ipc4;
+	};
```
Comment (ujfalusi) [inline include/uapi/sound/sof/header.h]: we might not need this change at all?
Leaving the `type` as it were and document that it is param_id in IPC4.
Looking at the firmware code, the type is used for the same purpose as the param_id.

**Comment by ujfalusi** [comment]
Changes since v6:
- Drop the modification of sof_abi_hdr, just document what the `type` is used in case of IPC4. At the end both IPC3 and IPC4 uses it for the same purpose, IPC3 sends it via IPC payload, IPC4 sends it via the header (param_id) and used as an ID of the blob to be used for.

**Review by ranj063 — APPROVED** [review]

**Review by juimonen — APPROVED** [review]
I'm fine with the abi_hdr change

File: sound/soc/sof/ipc4-control.c [hunk 13]
```diff
@@ -181,32 +181,262 @@ static int sof_ipc4_volume_get(struct snd_sof_control *scontrol,
...
+
+	/* make sure TLV info is consistent */
+	if (header.length + sizeof(struct snd_ctl_tlv) > size) {
+		dev_err_ratelimited(scomp->dev,
+				    "Inconsistent TLV, data %d + header %zu > %d\n",
+				    header.length, sizeof(struct snd_ctl_tlv), size);
+		return -EINVAL;
+	}
+
+	/* be->max is coming from topology */
```
Comment (kv2019i) [inline sound/soc/sof/ipc4-control.c:305]: Btw, this comment is also in IPC3 byte contorl implementation, but a bit hard to follow as  "be->max" is not used. "scontrol->max_size" is set based on "soc_bytes_ext", so I think this makes sense, but a bit hard to follow.

**Review by kv2019i — APPROVED** [review]
One minor "comment on comments", but this is not specific to this PR, so nothing blocking merge.

File: sound/soc/sof/ipc4-control.c (hunk 13 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-control.c:305]: I will do a followup PR to update both IPC3 and IPC4, is that OK?

File: include/uapi/sound/sof/header.h [hunk 14]
```diff
@@ -13,8 +13,12 @@
...
- * @magic: Magic number for validation: 0x00464F53 ('S', 'O', 'F', '\0')
- * @type: Component specific type
+ * @magic: Magic number for validation
+ *	   for IPC3 data: 0x00464F53 ('S', 'O', 'F', '\0')
+ *	   for IPC4 data: 0x34464F53 ('S', 'O', 'F', '4')
+ * @type: module specific parameter
+ *	  for IPC3: Component specific type
+ *	  for IPC4: parameter ID (param_id) of the data
  * @size: The size in bytes of the data, excluding this struct
  * @abi: SOF ABI version
```
Comment (plbossart) [inline include/uapi/sound/sof/header.h:23]: nit-pick: we should clarify that the ABI version is within the scope of the magic number, i.e. there is no relationship between the abi value for IPC3 and for IPC4.

**Review by plbossart — APPROVED** [review]

**Comment by ranj063** [comment]
Mergin this so as to not lose approvals. @ujfalusi please submit fixups if needed

File: include/uapi/sound/sof/header.h (hunk 14 above)
Comment (ujfalusi) [reply include/uapi/sound/sof/header.h:23]: Yes, it is a good idea, I will create a fixup for this.
